### PR TITLE
Code push additions

### PR DESCRIPTION
**Pull request by max-mironov**
The goal of this PR is to add CodePush `Promote/Rollback/Clear/Remove/Rename` commands to Mobile Center CLI.
Also it includes changes from https://github.com/Microsoft/mobile-center-cli/pull/288/files and Sergey's changes for `Release` Command (https://github.com/sergey-akhalkov/mobile-center-cli/commits/codepush/release). This merge was needed because these commands share some common functionality.

### CODE REVIEW

File: src/commands/codepush/lib/hash-utils.ts [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+    const CODEPUSH_METADATA = ".codepushrelease";
+    return startsWith(relativeFilePath, __MACOSX)
+      || relativeFilePath === DS_STORE
+      || endsWith(relativeFilePath, "/" + DS_STORE)
+      || relativeFilePath === CODEPUSH_METADATA
+      || endsWith(relativeFilePath, "/" + CODEPUSH_METADATA);
+  }
+}
+
+function startsWith(str: string, prefix: string): boolean {
```
Comment (christav) [inline src/commands/codepush/lib/hash-utils.ts]: Use `_.startsWith` from lodash instead of custom implementation.

File: src/commands/codepush/lib/hash-utils.ts [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  return str && str.substring(0, prefix.length) === prefix;
+}
+
+function endsWith(str: string, suffix: string): boolean {
```
Comment (christav) [inline src/commands/codepush/lib/hash-utils.ts]: Use `_.endsWith` from lodash instead of custom implementation.

File: src/commands/codepush/lib/update-contents-tasks/sign.ts [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+  };
+
+  return new Promise<void>((resolve, reject) => {
+    jwt.sign(claims, privateKey, { algorithm: "RS256" }, async (err: Error, signedJwt: string) => {
+      if (err) {
+        reject(new Error("The specified signing key file was not valid"));
+      }
+
+      try {
+        await pfs.writeFile(signatureFilePath, signedJwt);
```
Comment (christav) [inline src/commands/codepush/lib/update-contents-tasks/sign.ts]: Any reason you're not just using fs.writeFileSync here? We're not on a server, we can use sync file writes.

File: src/commands/codepush/lib/update-contents-tasks/zip.ts [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+import * as path from "path";
+import * as JsZip from "jszip";
+import { generateRandomFilename } from "../file-utils";
+
+interface ReleaseFile {
+  sourceLocation: string;     // The current location of the file on disk
+  targetLocation: string;     // The desired location of the file within the zip
+}
+
+export default async function zip(updateContentsPath: string): Promise<string> {
```
Comment (christav) [inline src/commands/codepush/lib/update-contents-tasks/zip.ts]: redundant async

File: src/commands/codepush/lib/update-contents-tasks/zip.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      zip.file(releaseFile.targetLocation, await pfs.readFile(releaseFile.sourceLocation));
+    });
+
+    zip
+      .generateNodeStream({ type: 'nodebuffer', streamFiles: true })
+      .pipe(fs.createWriteStream(packagePath))
+      .on('finish', function () {
+        resolve(packagePath);
+      });
+  })
```
Comment (christav) [inline src/commands/codepush/lib/update-contents-tasks/zip.ts]: Missing semicolon

File: src/commands/setup-autocomplete.ts [hunk 6]
```diff
@@ -10,7 +10,7 @@ import { out } from "../util/interaction";
 const debug = require("debug")("mobile-center-cli:commands:setup-autocomplete");
 import { inspect } from "util";
 
-@help("Setup tab completion for your shell")
+@help("Set up autocompletion for the shell")
```
Comment (christav) [inline src/commands/setup-autocomplete.ts]: These changes are unrelated, please send them in a separate PR.

File: src/util/misc/jszip-helper.ts [hunk 7]
```diff
@@ -51,8 +50,8 @@ export async function addFolderToZipRecursively(path: string, zip: JsZip): Promi
     if (subEntityStats.isDirectory()) {
       await addFolderToZipRecursively(subEntityPath, folderZip);
     } else {
-      const fileStream = Fs.createReadStream(subEntityPath);
-      folderZip.file(subEntityName, fileStream);
+      const fileBuffer = await Pfs.readFile(subEntityPath);
```
Comment (christav) [inline src/util/misc/jszip-helper.ts]: This changes the logic so that it reads the entire file into memory before zipping. This limits the size of files we can add to available RAM. Don't do this unless there's a really good reason.

File: src/commands/build/download.ts [hunk 8]
```diff
@@ -5,7 +5,7 @@ import { inspect } from "util";
 import * as _ from "lodash";
 import * as Process from "process";
 import * as Request from "request";
-import * as JsZip from "jszip";
+import * as JSZip from "jszip";
```
[the remark this answers is not in the document]
Comment (sergey-akhalkov) [reply src/commands/build/download.ts]: @christav, fixed, please review https://github.com/Microsoft/mobile-center-cli/pull/288

File: src/commands/codepush/lib/hash-utils.ts [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+export async function generatePackageHashFromDirectory(directoryPath: string, basePath: string): Promise<string> {
+  if (!(await pfs.stat(directoryPath)).isDirectory()) {
+    throw new Error("Not a directory. Please either create a directory, or use hashFile().");
+  }
+
+  let manifest: PackageManifest = await generatePackageManifestFromDirectory(directoryPath, basePath);
+  return await manifest.computePackageHash();
+}
+
+export async function generatePackageManifestFromDirectory(directoryPath: string, basePath: string): Promise<PackageManifest> {
```
[the remark this answers is not in the document]
Comment (sergey-akhalkov) [reply src/commands/codepush/lib/hash-utils.ts]: fixed

File: src/commands/codepush/lib/hash-utils.ts [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+                fileHashesMap.set(relativePath, hash);
+              });
+          }
+        });
+    }, Promise.resolve(<void>null));
+
+    generateManifestPromise
+      .then(() => {
+        resolve(new PackageManifest(fileHashesMap));
+      }, reject)
```
[the remark this answers is not in the document]
Comment (sergey-akhalkov) [reply src/commands/codepush/lib/hash-utils.ts]: fixed

File: src/util/misc/jszip-helper.ts (hunk 7 above)
Comment (sergey-akhalkov) [reply src/util/misc/jszip-helper.ts]: fixed

File: src/commands/codepush/lib/update-contents-tasks/zip.ts (hunk 5 above)
Comment (sergey-akhalkov) [reply src/commands/codepush/lib/update-contents-tasks/zip.ts]: fixed

File: src/commands/codepush/lib/update-contents-tasks/zip.ts (hunk 4 above)
Comment (sergey-akhalkov) [reply src/commands/codepush/lib/update-contents-tasks/zip.ts]: fixed

File: src/commands/codepush/lib/update-contents-tasks/sign.ts (hunk 3 above)
Comment (sergey-akhalkov) [reply src/commands/codepush/lib/update-contents-tasks/sign.ts]: fixed

File: src/commands/codepush/lib/hash-utils.ts (hunk 2 above)
Comment (sergey-akhalkov) [reply src/commands/codepush/lib/hash-utils.ts]: fixed

File: src/commands/codepush/lib/hash-utils.ts (hunk 1 above)
Comment (sergey-akhalkov) [reply src/commands/codepush/lib/hash-utils.ts]: fixed

File: src/commands/codepush/lib/hash-utils.ts [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+    var entries: string[] = [];
+    this._map.forEach((hash: string, name: string): void => {
+      entries.push(name + ":" + hash);
+    });
+
+    // Make sure this list is alphabetically ordered so that other clients
+    // can also compute this hash easily given the update contents.
+    entries = entries.sort();
+
+    return Promise.resolve(
```
[the remark this answers is not in the document]
Comment (sergey-akhalkov) [reply src/commands/codepush/lib/hash-utils.ts]: fixed

File: src/commands/codepush/lib/hash-utils.ts [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        hashStream.end();
+
+        var buffer = <Buffer>hashStream.read();
+        var hash: string = buffer.toString("hex");
+
+        resolve(hash);
+      });
+
+    readStream.pipe(hashStream);
+  })
```
[the remark this answers is not in the document]
Comment (sergey-akhalkov) [reply src/commands/codepush/lib/hash-utils.ts]: fixed

File: src/commands/codepush/lib/hash-utils.ts [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+      }, reject)
+  })
+}
+
+export async function hashFile(filePath: string): Promise<string> {
+  var readStream: fs.ReadStream = fs.createReadStream(filePath);
+  return hashStream(readStream);
+}
+
+export async function hashStream(readStream: stream.Readable): Promise<string> {
```
[the remark this answers is not in the document]
Comment (sergey-akhalkov) [reply src/commands/codepush/lib/hash-utils.ts]: fixed

File: src/commands/codepush/lib/hash-utils.ts [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+    }, Promise.resolve(<void>null));
+
+    generateManifestPromise
+      .then(() => {
+        resolve(new PackageManifest(fileHashesMap));
+      }, reject)
+  })
+}
+
+export async function hashFile(filePath: string): Promise<string> {
```
[the remark this answers is not in the document]
Comment (sergey-akhalkov) [reply src/commands/codepush/lib/hash-utils.ts]: fixed

File: src/commands/setup-autocomplete.ts (hunk 6 above)
Comment (max-mironov) [reply src/commands/setup-autocomplete.ts]: Yep, agreed, that was accidentally picked up from this PR: https://github.com/Microsoft/mobile-center-cli/pull/250, I'll revert it. Please let me know if these changes ever needed so I can create another PR for this.

File: src/commands/codepush/deployment/add.ts [hunk 15]
```diff
@@ -26,19 +26,19 @@ export default class AddCommand extends AppCommand {
...
         (cb) => client.codePushDeployments.create(app.ownerName, app.appName, this.newDeploymentName, cb)));
       deployment = httpRequest.result;
-      out.text(`Deployment ${deployment.name} has been created for the app ${app.appName}`);
+      out.text(`Deployment ${chalk.bold(deployment.name)} has been created for the ${this.identifier}`);
       return success();
     } catch (error) {
       debug(`Failed to add a new CodePush deployment - ${inspect(error)}`);
       if (error.statusCode === 404) {
-        const appNotFoundErrorMsg = `The app ${app.ownerName}/${app.appName} does not exist. Please double check the name, and provide it in the form owner/appname. \nRun the command ${chalk.bold("mobile-center apps list")} to see what apps you have access to.`;
+        const appNotFoundErrorMsg = `The app ${this.identifier} does not exist. Please double check the name, and provide it in the form owner/appname. \nRun the command ${chalk.bold("mobile-center apps list")} to see what apps you have access to.`;
```
[the remark this answers is not in the document]
Comment (max-mironov) [reply src/commands/codepush/deployment/add.ts]: Thanks for pointing, fixed!
